Enable import for Folder/Study archives when paired with the cloud module

Sigmonia opened the pull request:
#### Rationale
Item #9160 - [Labkey R&D-EI - Reload Study from S3 root via Folder Management UI](https://docs.google.com/document/d/1Fe9tyqIdvMTyu7gFYm4PJrfFMXMf1oQbpVXXdvRzqn8/)

#### Related Pull Requests
* https://github.com/LabKey/cloud/pull/87
* https://github.com/LabKey/dataintegration/pull/122
* https://github.com/LabKey/testAutomation/pull/847
* https://github.com/LabKey/commonAssays/pull/360
* https://github.com/LabKey/targetedms/pull/423
* https://github.com/LabKey/immport/pull/42
* https://github.com/LabKey/moduleEditor/pull/45

#### Changes
* Refactors to replace some usages of File with nio.Path
  * Deprecated several API methods in preference for Path based versions
  * Fixed some typos & NPEs along the way
* Added cloud service task to download study archive files
  * Adjusted context and root Job properties to account for cloud --> temp directory path change during processing

Sigmonia committed:
Checkpoint: Study import from zip
- Mostly swapped File --> nio.Path and subsequent refactors

TODO: Lots...

Sigmonia committed:
Remove root check

Sigmonia committed:
Fix usage of makeLegalName
Fix error with importDir Path being changed

Sigmonia committed:
*in-progress* - builds, Expanded archive Files download. Query loading fails because it is still looking at the s3 root

TODO: Move download to PipelineManager.getImortXmlFile -- Hate the sideaffecting location, but I'm not sure how else to map the ImportContextRoot to the importDir

Sigmonia committed:
*in-progress* - Load secondary files (query, vists, etc.)

TODO: needs cleanup and better edge case handling.

Sigmonia committed:
*in-progress* - more targeted cloud support

Sigmonia committed:
Fix study import not being enabled
clean up some comment lint

Sigmonia committed:
Fix specimen files not getting listed & imported correctly

Sigmonia committed:
Fix StudyContinuousTest
clean up some lint warnings in FileSystemFile

Sigmonia committed:
Fix Path endsWith issue

Sigmonia committed:
Fix compilation issue for biologics

Sigmonia committed:
Fix NPE

Sigmonia committed:
Fix logging error

Sigmonia committed:
Fix issue parsing BOM

Sigmonia committed:
Push Job log to s3 location

Sigmonia committed:
Move download process from job creation to job execution
- removed final from job ImportContext and file root properties. Download requires updating those values to the download locations after job creation.
Add server-level Cloud studies count metric 'cloudBackedStudies'

Fixed list filename logging typo
Fixed issue where lists with spaces weren't being loaded/copied correctly
Throw exception when cloud service isn't present and attempting import of remote archives

Sigmonia committed:
Restore and Deprecate the File version of the runStudyImportJob

Sigmonia committed:
Fix NPE when importing folder.zip

Sigmonia committed:
Fix error with the notifications task's logging that failed for cloud jobs
Enabled cloud FolderImport for *.folder.zip

TODO - enable cloud for expanded folder archives

Sigmonia committed:
Enable cloud import for expanded folder archives

TODO: Could consolidate the Folder/Study archive downloads...

labkey-klum commented on api/src/org/labkey/api/admin/FolderImportContext.java:
```diff
@@ -83,44 +104,40 @@ public synchronized FolderDocument getDocument() throws ImportException, Invalid
-    }
-
-    public File getFolderFile(File root, File dir, String name, String source) throws ImportException
+//    // Assume file was referenced in folder.xml file   // TODO: Context should hold onto the root -- shouldn't have to pass it in
```
is there a reason to keep this code around?

Sigmonia replied:
removed

labkey-klum commented on api/src/org/labkey/api/pipeline/PathSerialization.java:
```diff
@@ -71,14 +74,24 @@ protected Deserializer()
         public Path deserialize(JsonParser parser, DeserializationContext context) throws IOException
```
add a `Nullable` annotation?

Sigmonia replied:
added

labkey-klum commented on api/src/org/labkey/api/pipeline/PathSerialization.java:
```diff
@@ -71,14 +74,24 @@ protected Deserializer()
+                    //TODO need LKS container to pull the appropriate S3 credentials/FileSystem representation (in the event of interrupted job)
+                }
+
+                return Path.of(str); //TODO this will cause issues...
```
would it be better to throw then?

Sigmonia replied:
If the string path isn't viable it will cause an exception, the existing message is fairly clear (at least as clear as I would make it) rather than catch and rethrow, I think it makes sense to leave it as is and let the parser throw if needed.

labkey-klum commented on api/src/org/labkey/api/pipeline/PipelineProvider.java:
```diff
@@ -249,7 +249,7 @@ public Module getOwningModule()
     public boolean supportsCloud()
     {
-        return false;
+        return true;
```
Are we really saying that all providers now support cloud storage? Wouldn't it be safer to keep this unchanged and just override.

Sigmonia replied:
switched

labkey-klum commented on api/src/org/labkey/api/study/StudyService.java:
```diff
@@ -179,7 +180,13 @@ static StudyService get()
 
-    boolean runStudyImportJob(Container c, User user, @Nullable ActionURL url, File studyXml, String originalFilename, BindException errors, PipeRoot pipelineRoot, ImportOptions options);
+    @Deprecated //prefer the Path version
+    default boolean runStudyImportJob(Container c, User user, @Nullable ActionURL url, File studyXml, String originalFilename, BindException errors, PipeRoot pipelineRoot, ImportOptions options)
```
Is there any code that references this method?

Sigmonia replied:
I'm a little concerned an external reference may exist, but I removed the method for now.

labkey-klum commented on api/src/org/labkey/api/util/FileType.java:
```diff
@@ -605,16 +638,35 @@ public String toString()
     }
 
+//    @NotNull
+//    public static List<FileType> findTypes(@NotNull List<FileType> types, @NotNull List<File> files)
```
delete?

Sigmonia replied:
deleted

labkey-susanh commented on api/src/org/labkey/api/admin/InvalidFileException.java:
```diff
@@ -22,14 +22,21 @@
 
 public class InvalidFileException extends ImportException
 {
+    @Deprecated // prefer the Path version
```
Would we want to have this forward to the Path version?
```
this(root.toPath(), file.toPath(), t);
```

Sigmonia replied:
fixed

labkey-klum commented on study/src/org/labkey/study/pipeline/StudyReloadSourceJob.java:
```diff
@@ -141,4 +152,46 @@ public boolean isMerge()
     }
+
+    @Override
+    public void downloadCloudArchive(Path studyXml, BindException errors)
```
This method is nearly identical to the version in `StudyImportJob`, we should try to consolidate otherwise it will probably bite us down the road.

Sigmonia replied:
consolidated

labkey-susanh commented on api/src/org/labkey/api/pipeline/PathSerialization.java:
```diff
@@ -71,14 +74,24 @@ protected Deserializer()
         public Path deserialize(JsonParser parser, DeserializationContext context) throws IOException
         {
             String str = parser.getValueAsString();
+//            return Path.of(URI.create(str));
```
remove commented code?

Sigmonia replied:
removed

labkey-susanh commented on api/src/org/labkey/api/pipeline/PathSerialization.java:
```diff
@@ -71,14 +74,24 @@ protected Deserializer()
+                    //TODO this will likely work only for cloud paths that include access ids eg: s3://<pub-access-key>@s3.amazonaws.com/<my bucket>/...
+                    return css.getPathFromUrl(str);
+                    //TODO need LKS container to pull the appropriate S3 credentials/FileSystem representation (in the event of interrupted job)
+                }
```
All of these TODOs make me wonder if this does work.  If it works in limited cases, do we need or have some corresponding documentation?

labkey-susanh commented on api/src/org/labkey/api/pipeline/PipeRoot.java:
```diff
@@ -69,6 +69,20 @@
     @Nullable
     File resolvePath(String relativePath);
 
+    //TODO implement this properly
```
Properly?  Do we know what that means here?  Something to do now?

Sigmonia replied:
this TODO is a left over implementation thought.

I think using the implemented getRootNioPath will work correctly and gives the Root Impl the authority of how to relativize the Path.

labkey-susanh commented on api/src/org/labkey/api/pipeline/PipelineJob.java:
```diff
@@ -123,6 +123,11 @@ public void clearActionSet(ExpRun run)
 
+    public void updateWorkingRoot(Path importRoot) throws PipelineJobException
+    {
+        throw new PipelineJobException("Unable to update working root. Updates not supported by job");
```
```suggestion
        throw new PipelineJobException("Unable to update working root. Updates not supported by job.");

```

Sigmonia replied:
added

labkey-klum commented on study/src/org/labkey/study/importer/StudyImportInitialTask.java:
```diff
@@ -62,6 +65,13 @@ public RecordedActionSet run() throws PipelineJobException
         StudyImportContext ctx = support.getImportContext();
 
+        boolean isCloudRoot = job.getPipeRoot().isCloudRoot();
+        if (isCloudRoot)
```
`StudyFolderImporter` also checks whether the study dir has a cloud scheme. Should we do the same here?

Sigmonia replied:
yes, I think...

labkey-susanh commented on api/src/org/labkey/api/pipeline/PipelineJob.java:
```diff
@@ -1086,7 +1091,7 @@ public void run()
     protected void finallyCleanUpLocalDirectory()
     {
-        if (null != _localDirectory)
+        if (null != _localDirectory & isDone())
```
when is this "finallyCleanUp" called when the job is not done?  Is this protection against an error case?  Or a case when nothing happened in the first place?

Sigmonia replied:
This is protection against an error case -- specifically when there is an error and there are notifications to clean up. See line ~1813 where this method is called. The method references the log file, but if this method executes before that then the log file has moved--so it causes a secondary FileNotFoundException.

labkey-susanh commented on api/src/org/labkey/api/pipeline/file/FileAnalysisJobSupport.java:
```diff
@@ -76,6 +79,25 @@
 
+    default Path findInputPath(String filepath)
+    {
+        //TODO This needs implementation in derived classes...
```
Is this to be done in this PR?

Sigmonia replied:
Short answer is no. Some of these are complicated, and also not part of LabKey's managed code

labkey-susanh commented on api/src/org/labkey/api/writer/FileSystemFile.java:
```diff
@@ -55,31 +56,42 @@ private FileSystemFile()
+        }
+        catch (IOException e)
+        {
+            //TODO do more...
```
More?  What did you have in mind?

Sigmonia replied:
Was thinking some additional logging, but there isn't an obvious logger available.

Removed.

labkey-susanh commented on api/src/org/labkey/api/writer/FileSystemFile.java:
```diff
@@ -162,42 +173,63 @@ public XmlObject getXmlBean(String filename) throws IOException
+        }
+        catch (IOException e)
+        {
+            //TODO should do more here...
```
Do more?  Is there some way to recover?

Sigmonia replied:
Was thinking some additional logging, but there isn't an obvious logger available.

Removed.

labkey-susanh commented on api/src/org/labkey/api/writer/FileSystemFile.java:
```diff
@@ -162,42 +173,63 @@ public XmlObject getXmlBean(String filename) throws IOException
+        }
+        catch (IOException e)
+        {
+            //TODO do more here...
```
Same comment here.  Let's either remove this TODO or make the comment more meaningful or do whatever "more" is in this case.

Sigmonia replied:
Was thinking some additional logging, but there isn't an obvious logger available.

Removed

labkey-susanh commented on api/src/org/labkey/api/writer/PrintWriters.java:
```diff
@@ -51,7 +54,31 @@ public static PrintWriter getPrintWriter(OutputStream out)
+        }
+        catch (FileNotFoundException e)
+        {
+            // TODO Retain old behavior
```
What is the old behavior that needs to be retained?  I'm guessing we don't want to change the behavior.

Sigmonia replied:
It was just throwing, removed TODO, and move try block to Path method to catch the new IOException.

labkey-susanh commented on experiment/src/org/labkey/experiment/api/ExpDataTableImpl.java:
```diff
@@ -937,7 +938,7 @@ public void testWebDavColumns() throws Exception
                 {
                     VirtualFile f = new MemoryVirtualFile();
-                    excel.write(f.getOutputStream("excel.xlsx"));
+                    excel.write(f.getOutputStream("excel.xlsx")); //TODO should this outputStream be closed?
```
Closing the stream seems a good idea unless that happens elsewhere.

Sigmonia replied:
done.

labkey-susanh commented on pipeline/src/org/labkey/pipeline/PipelineDirectoryImpl.java:
```diff
@@ -32,6 +32,7 @@
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Stream;
```
Needed?

Sigmonia replied:
removed

labkey-susanh commented on pipeline/src/org/labkey/pipeline/api/PipeRootImpl.java:
```diff
@@ -413,7 +420,9 @@ public File getImportDirectory()
         return new File(root, PipelineService.UNZIP_DIR);
     }
 
+    //TODO why do we have both of these? 1
```
Are both referenced?

Sigmonia replied:
not any more...

labkey-susanh commented on pipeline/src/org/labkey/pipeline/api/PipelineManager.java:
```diff
@@ -805,68 +814,100 @@ else if (!pipeRoot.isCloudRoot() && !pipeRoot.isUnderRoot(archiveFile))     // T
             }
         }
 
+        return importDir.toAbsolutePath().resolve(xmlFileName);
```
Can `xmlFileName` be null?

Sigmonia replied:
Not as far as I can tell, I trace it back to static strings from PipelineController:~1319

labkey-susanh commented on pipeline/src/org/labkey/pipeline/importer/FolderImportTask.java:
```diff
@@ -137,6 +145,11 @@ public RecordedActionSet run() throws PipelineJobException
 
+    private boolean hasCloudRoot(PipelineJob job)
+    {
+        return job.getPipeRoot().isCloudRoot(); //TODO should this check to target file's protocol?
```
I don't know the answer to the TODO question.  Is there a way to do that check here?  How would we have `isCloudRoot` return true but the protocol is not as expected?

Sigmonia replied:
Consolidated to use the FileUtil to check URI protocol.  (My thought was if we ever allowed GCS) I think by using the helper method here we can ensure that that would be the case.

labkey-susanh commented on study/api-src/org/labkey/api/specimen/pipeline/AbstractSpecimenTask.java:
```diff
@@ -188,7 +198,12 @@ public ImportHelper getImportHelper()
+//        default VirtualFile getSpecimenDir(PipelineJob job, SimpleStudyImportContext ctx, @Nullable File inputFile) throws IOException, ImportException, PipelineJobException
+//        {
+//            return getSpecimenDir(job, ctx, inputFile !=  null ? inputFile.toPath() : null);
+//        }
```
Remove?

Sigmonia replied:
done.

labkey-susanh commented on study/src/org/labkey/study/importer/StudyImportJob.java:
```diff
@@ -96,6 +96,65 @@ public StudyImportJob(Container c, User user, ActionURL url, File studyXml, Stri
+            {
+                try
+                {
+                    Path importRoot = CloudStoreService.get().copyExpandedStudyArchiveLocally(studyXml, _ctx.getXml(), _ctx.getRoot().getLocation(), getPipeRoot(), _ctx.getLoggerGetter().getLogger(), errors);
```
Probably a naive question at this stage, but is there a chance that the study in the cloud is going to be too big to copy locally?

Sigmonia replied:
Absolutely it is possible. Currently jClouds would be the generator of the Exception on the download attempt. We don't currently check for size constraints but we probably should. I think this might be worth breaking out as a bug to follow up with after the main story is solidified. We also dont do this for the TargetedMS imports AFICT

labkey-susanh commented on study/src/org/labkey/study/importer/StudyImporterFactory.java:
```diff
@@ -104,14 +106,22 @@ public void process(PipelineJob job, ImportContext<FolderDocument.Folder> ctx, V
                 User user = ctx.getUser();
                 BindException errors = new NullSafeBindException(c, "import");
 
+                //TODO unify with other download location
```
Is this to be done now?

Sigmonia replied:
done.

labkey-susanh commented on study/src/org/labkey/study/importer/StudyImporterFactory.java:
```diff
@@ -121,7 +131,14 @@ public void process(PipelineJob job, ImportContext<FolderDocument.Folder> ctx, V
+                        .withRoot(useLocalImportDir ? new FileSystemFile(job.getPipeRoot().getImportDirectory()) : studyDir)
+                        .build();
+
+//                new StudyImportContext(user, c, studyDoc, ctx.getDataTypes(), ctx.getLoggerGetter(), studyDir);
```
remove?

Sigmonia replied:
removed

labkey-susanh commented on study/src/org/labkey/study/importer/StudyReload.java:
```diff
@@ -178,13 +182,24 @@ public ReloadStatus attemptReload(ImportOptions options, String source) throws I
+            {
+                throw new UnauthorizedException("Reload user is inactive");
+            }
+            if (!study.hasPermission(user, UpdatePermission.class))  //TODO double check required permission
```
The permission seems correct to me, but you could verify by impersonating an editor and seeing if reload is available.

Sigmonia replied:
verified.

labkey-susanh commented on study/src/org/labkey/study/importer/StudyReload.java:
```diff
@@ -202,36 +217,27 @@ public ReloadStatus reloadStudy(StudyImpl study, ImportOptions options, String s
                 else
                 {
-                    studyXml = root.resolvePath("study.xml");
+                    studyXml = root.resolveRelativePath("study.xml");  //TODO this is likely wrong...
```
Do we know what the right answer is or why it is likely wrong?  Would be worth adding a comment about why it seems wrong, at least.

Sigmonia replied:
The thought was that I was using the incorrect resolution method, but I think it is actually correct. Looking at the previous method AFICT they resolve the same.

labkey-klum commented on pipeline/src/org/labkey/pipeline/importer/FolderImportTask.java:
```diff
@@ -59,15 +61,21 @@ public RecordedActionSet run() throws PipelineJobException
 
+        if (hasCloudRoot(job) && CloudStoreService.get() != null)
+        {
+            Path importRoot = CloudStoreService.get().downloadFolderArchive(job);
```
I'm confused by the abstraction of the code which handles the downloading of bits from a cloud store. In this case we are interfacing directly with the `CloudStoreService` to download the archive. In the study import and reload cases the pipeline jobs which implement `StudyJobSupport.downloadCloudArchive` are called on to handle the download. Is there a reason why the pipeline job in the folder import case can't do the same thing?

Additionally the code in `CloudFolderImporter` and `CloudStudyImporter` could be pushed together as they already share some duplicate code. It seems like the only distinction is one downloads a folder archive and the other downloads an expanded archive?

labkey-klum replied:
If we did push more of this code into the pipeline job then `updateWorkingRoot` could be handled inside of the job as part of the download operation. Then we don't need to add the method to the abstract base since it seems like calling this method only makes sense in the context of downloading into a temp location (the study pipeline jobs seem to be doing this in `generateImportContext`).

Sigmonia replied:
So I did unify the download code into the `CloudArchiveImportSupport` which is in turn implemented by the jobs. This is also where the updateWorkingRoot method is now implemented. I think this allows us to isolate the cloud support to those jobs that implement that job supporter.

Sigmonia committed:
Fix zip archive imports
TODO: Could consolidate the Folder/Study archive downloads...

labkey-klum reviewed (changes requested):
It would be nice to clean up the abstraction of the code which invokes and implements the downloading of either a compressed or expanded study or folder archive.

Sigmonia committed:
Some initial cleanup post CR
- remove commented out code
- change default value for supportsCloud method

Sigmonia committed:
remove commented out code

Sigmonia committed:
Consolidate study jobs download of archive

Sigmonia committed:
CR Feedback
- Route file based InvalidFileException constructor through the Path constructor
- Typo & grammar fixes
- Clean-up some TODOs
- Unify method to check for whether to download archive

Sigmonia committed:
clean up some lint markers

Sigmonia committed:
Consolidate expanded Study and Folder archives to one download importer.

Sigmonia committed:
Remove TODOs

Sigmonia committed:
Added some additional comments regarding TODOs

Sigmonia committed:
Remove TODOs and redundant catch

Sigmonia committed:
Close output stream

Sigmonia committed:
Removed some redundant methods that are no longer referenced

Sigmonia committed:
clean-up a commented out method

Sigmonia committed:
Fix job support conversion break due to previous change

labkey-jeckels commented on pipeline/src/org/labkey/pipeline/api/PipelineManager.java:
```diff
@@ -805,68 +814,94 @@ else if (!pipeRoot.isCloudRoot() && !pipeRoot.isUnderRoot(archiveFile))     // T
+    public static @NotNull Path getXmlFilePathFromArchive(Path importDir, Path archiveFile, @NotNull String xmlFileName) throws InvalidFileException
+    {
+        // when importing a folder archive for a study, the study.xml file may not be at the root
+        if ("study.xml".equals(xmlFileName) && archiveFile.getFileName().toString().endsWith(".folder.zip"))
```
Case sensitivity

Sigmonia replied:
fixed

labkey-jeckels commented on pipeline/src/org/labkey/pipeline/api/PipelineManager.java:
```diff
@@ -805,68 +814,94 @@ else if (!pipeRoot.isCloudRoot() && !pipeRoot.isUnderRoot(archiveFile))     // T
+    {
+        Path xmlFile = archiveFile;
+
+        if (archiveFile.getFileName().toString().endsWith(".zip"))
```
Case sensititivity

Sigmonia replied:
fixed

labkey-jeckels commented on study/api-src/org/labkey/api/specimen/pipeline/AbstractSpecimenTask.java:
```diff
@@ -197,22 +207,22 @@ public ImportHelper getImportHelper()
                             job.setStatus("TRANSFORMING " + transformer.getName() + " DATA");
-                        specimenArchive = ARCHIVE_FILE_TYPE.getFile(inputFile.getParentFile(), transformer.getFileType().getBaseName(inputFile));
-                        transformer.transform(job, inputFile, specimenArchive);
+                        specimenArchive = ARCHIVE_FILE_TYPE.getFile(inputFile.getParent().toFile(), transformer.getFileType().getBaseName(inputFile));
```
Won't this blow up if it's S3-backed?

Sigmonia replied:
Made a change here to use nio.Path

labkey-jeckels commented on study/api-src/org/labkey/api/specimen/pipeline/AbstractSpecimenTask.java:
```diff
@@ -250,18 +260,38 @@ public void afterImport(SimpleStudyImportContext ctx)
+                                }
+                                catch (IOException e)
+                                {
+                                    log.error("Unable to delete file: " + p);
```
Include exception here like a few lines lower?

Sigmonia replied:
Adjusted to use FileUtil.dirDelete

labkey-jeckels commented on study/src/org/labkey/study/importer/StudyImportProvider.java:
```diff
@@ -55,12 +58,21 @@ public void updateFileProperties(ViewContext context, PipeRoot pr, PipelineDirec
+            // Path.endsWith checks the final path piece (eg, /root/filename.xml checks filename.xml).
+            // So, we want to check the string value instead
+            String pathString = entry.toString();
+            return pathString.endsWith("study.xml") || pathString.endsWith(".study.zip") || pathString.endsWith(".folder.zip");
```
Case sensitivity

Sigmonia replied:
fixed

labkey-jeckels reviewed (commented):
Smaller suggestions.

Also, there aren't a lot of places here where we check if CloudStoreService.get() == null, but eventually we'll probably want a dummy implementation that lets us avoid doing null checks everywhere, throwing UnsupportedOperationExceptions or no-opping as appropriate.

Sigmonia committed:
Fix potential casing check issue

Sigmonia committed:
Fell down the rabbit hole of replacing File with nio.Path. Had to cut it short eventually, so still too many calls to toFile/toPath in order to fill gaps.
Update ShowFileAction to stream Paths
-- Fixes Log file link for Cloud jobs

Sigmonia committed:
fix a case sensitivity check for the standardized import files
Use utility method for performing the temp directory delete

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_s3StudyImport

Sigmonia committed:
Fix build issues with previous push

Sigmonia committed:
Clean up some lint
Move updateWorkingRoot method to the CloudArchiveImporterSupport interface

Sigmonia committed:
Revert deprecated methods back to original return types

Sigmonia committed:
Merge remote-tracking branch 'origin/develop' into fb_s3StudyImport

Sigmonia committed:
create log file for path if it doesn't exist

Sigmonia committed:
Don't look up context for cloud imports as it requires streaming the file
Fix some 'Raw use of Parameterized type' lint

Sigmonia committed:
Rollback some exception handling changes due to inheritance issues.

labkey-tchad committed:
Add missing unzipToDirectory

labkey-tchad committed:
Include submit button in calculation of panel height

labkey-tchad committed:
Merge remote-tracking branch 'origin/develop' into fb_s3StudyImport

Sigmonia committed:
Fix Path deserialization type mismatch

Sigmonia committed:
Merge remote-tracking branch 'origin/fb_s3StudyImport' into fb_s3StudyImport

Sigmonia committed:
Hide the "Apply to multiple folders" advanced import option for cloud based roots

Sigmonia committed:
Fix xar file naming issue after File to Path change

Sigmonia committed:
Fix file naming post Path refactor.

labkey-tchad committed:
Merge remote-tracking branch 'origin/develop' into fb_s3StudyImport

labkey-tchad committed:
Merge remote-tracking branch 'origin/develop' into fb_s3StudyImport

Sigmonia committed:
Hide the "Apply to multiple folders" advanced import option for cloud based roots

labkey-tchad committed:
Remove unnecessary parameter from folder import helper
